File: java/leetcode_grind/Day722.java
package leetcode_grind;

public class Day722 {
    // https://leetcode.com/problems/maximum-xor-for-each-query/description/?envType=daily-question&envId=2024-11-08
    static class Solution1 {
        public int[] getMaximumXor(int[] nums, int maximumBit) {
            int xorProduct = 0;
            for (int i = 0; i < nums.length; i++) {
                xorProduct = xorProduct ^ nums[i];
            }
            int[] ans = new int[nums.length];
            int mask = (1 << maximumBit) - 1;
            for (int i = 0; i < nums.length; i++) {
                ans[i] = xorProduct ^ mask;
                xorProduct = xorProduct ^ nums[nums.length - 1 - i];
            }
            return ans;
        }
    }
}

File: src/leetcode_grind/day_722.rs
// https://leetcode.com/problems/maximum-xor-for-each-query/description/?envType=daily-question&envId=2024-11-08
pub fn get_maximum_xor(nums: Vec<i32>, maximum_bit: i32) -> Vec<i32> {
    let mut prefix_xor = vec![0; nums.len()];
    prefix_xor[0] = nums[0];
    for i in 1..nums.len() {
        prefix_xor[i] = prefix_xor[i - 1] ^ nums[i];
    }
    let mut ans = vec![0; nums.len()];

    let mask = (1 << maximum_bit) - 1;
    for i in 0..nums.len() {
        let current_xor = prefix_xor[prefix_xor.len() - 1 - i];
        ans[i] = current_xor ^ mask;
    }
    ans
}
